util/dvsim/Scheduler.py

```diff
@@ -233,44 +104,372 @@ def on_sigint(signal_received, frame):
         finally:
             signal(SIGINT, old_handler)
 
-        # We got to the end without anything exploding. Return the results for our jobs.
+        # We got to the end without anything exploding. Return the results.
         return self.item_to_status
 
+    def _schedule_items(self, items):
+        '''Convert a heterogenous list of items to run into list of tasks
+        segregated by target and cfg (DUT).'''
 
-class Scheduler:
-    '''An object to run one or more Deploy items'''
+        # Scheduled items are ordered by targets.
+        scheduled = OrderedDict()
 
-    # Max jobs running at one time
-    max_parallel = 16
+        def add_to_scheduled(item):
+            '''Recursively go through the item's dependencies and itself to
+            the task list, associated with target and the cfg name.'''
 
-    # Max jobs dispatched in one go.
-    slot_limit = 20
+            for dep in item.dependencies:
+                add_to_scheduled(dep)
 
-    def __init__(self, items):
-        # An ordered dictionary keyed by target ('build', 'run' or similar).
-        # The value for each target is a TargetScheduler object.
-        self.schedulers = OrderedDict()
+            if item.target not in scheduled:
+                scheduled[item.target] = {}
+
+            if item.sim_cfg not in scheduled[item.target]:
+                scheduled[item.target][item.sim_cfg] = []
+
+            if item not in scheduled[item.target][item.sim_cfg]:
+                scheduled[item.target][item.sim_cfg].append(item)
```

Comment:
You can use `setdefault` to make this a bit simpler (and it will probably run faster too, not that we really care)
```python3
            for_tgt = scheduled.setdefault(item.target, [])
            for_cfg = for_tgt.setdefault(item.sim_cfg, [])
            if item not in for_cfg:
                for_cfg.append(item)
```